File: projetjee/src/main/java/com/examen/webservice/TaskService.java
package com.examen.webservice;

import java.util.List;

import com.examen.entities.Project;
import com.examen.entities.Task;
import com.examen.metier.interfaces.ProjectDao;
import com.examen.metier.interfaces.TaskDao;

import jakarta.ejb.EJB;
import jakarta.ejb.Stateless;
import jakarta.ws.rs.Consumes;
import jakarta.ws.rs.DELETE;
import jakarta.ws.rs.GET;
import jakarta.ws.rs.POST;
import jakarta.ws.rs.PUT;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.PathParam;
import jakarta.ws.rs.Produces;
import jakarta.ws.rs.core.MediaType;
import jakarta.ws.rs.core.Response;

@Stateless
@Path("/task")
public class TaskService {

	@EJB
	TaskDao taskDao;
	
	@EJB
	ProjectDao projectDao; 

	@GET
	@Path("/all")
	@Produces(MediaType.APPLICATION_JSON)
	public Response getAllTasks() {
	    try {
	        List<Task> tasks = taskDao.getAllTasks();
	        return Response.status(Response.Status.OK).entity(tasks).build();
	    } catch (Exception e) {
	        return Response.status(Response.Status.INTERNAL_SERVER_ERROR).entity("Error retrieving tasks").build();
	    }
	}

	@POST
	@Path("/create")
	@Consumes(MediaType.APPLICATION_JSON)
	@Produces(MediaType.APPLICATION_JSON)
	public Response createTask(Task task) {
	    try {
	        // Fetch the project from the database based on the provided code
	        Project project = projectDao.getProjectByCode(task.getProject().getCode());

	        // Create the task with the existing project
	        Task t = new Task(task.getCode(), task.getDescription(), task.getStartDate(), task.getEndDate(), project);
	        taskDao.createTask(t);

	        return Response.status(Response.Status.CREATED).entity(t).build();
	    } catch (Exception e) {
	        return Response.status(Response.Status.INTERNAL_SERVER_ERROR).entity("Error creating task").build();
	    }
	}



	@PUT
	@Path("/update")
	@Consumes(MediaType.APPLICATION_JSON)
	@Produces(MediaType.APPLICATION_JSON)
	public Response updateTask(Task task) {
	    try {
	    	Project project = projectDao.getProjectByCode(task.getProject().getCode());

	        Task t = new Task(task.getCode(), task.getDescription(), task.getStartDate(), task.getEndDate(), project);
	        taskDao.updateTask(t);
	        return Response.status(Response.Status.OK).entity(t).build();
	    } catch (Exception e) {
	        return Response.status(Response.Status.INTERNAL_SERVER_ERROR).entity("Error updating task").build();
	    }
	}

	@DELETE
	@Path("/remove/{code}")
	@Produces(MediaType.APPLICATION_JSON)
	public Response removeTask(@PathParam(value="code") String code) {
	    try {
	        Task removedTask = taskDao.removeTask(code);
	        if (removedTask != null) {
	            return Response.status(Response.Status.OK).entity(removedTask).build();
	        } else {
	            return Response.status(Response.Status.NOT_FOUND).entity("Task not found").build();
	        }
	    } catch (Exception e) {
	        return Response.status(Response.Status.INTERNAL_SERVER_ERROR).entity("Error removing task").build();
	    }
	}

}

File: projet_jee_frontend/src/services/taskService.js
// taskService.js
import axios from 'axios';

const TASK_BASE_URL = 'http://localhost:8080/Examen2/TaskService';

const taskService = {
  getAllTasks: async () => {
    try {
      const response = await axios.get(`${TASK_BASE_URL}/tasks`);
      return response.data;
    } catch (error) {
      console.error('Error fetching tasks:', error);
      return [];
    }
  },

  getTaskByCode: async (code) => {
    try {
      const response = await axios.get(`${TASK_BASE_URL}/tasks/${code}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching task details:', error);
      return {};
    }
  },

  addTask: async (task) => {
    try {
      const response = await axios.post(`${TASK_BASE_URL}/tasks/add`, task);
      return response.data;
    } catch (error) {
      console.error('Error adding task:', error);
      throw error;
    }
  },

  removeTask: async (code) => {
    try {
      const response = await axios.delete(`${TASK_BASE_URL}/tasks/delete/${code}`);
      return response.data;
    } catch (error) {
      console.error('Error removing task:', error);
      throw error;
    }
  },

  updateTask: async (task) => {
    try {
      const response = await axios.put(`${TASK_BASE_URL}/tasks/update/${task.code}`, task);
      return response.data;
    } catch (error) {
      console.error('Error updating task:', error);
      throw error;
    }
  },
};

export default taskService;
